### ISSUE TITLE
Background Modeling

### ISSUE THREAD

**Issue by grburgess** [giacomov/3ML#227]
I've been looking at the background modeling and it may be better to approach it by adding a node onto PointSource:

```python
ps = PointSource(...,background_shape=Powerlaw() )
```

Otherwise, we are just adding a function inside of the plugin and have to dig into private members to access the free parameters of that function. 

Is this too involved of an approach @giacomov ?

**Comment by volodymyrss** [comment]
Sorry to intervene, just I was dealing with similar situations in astromodels. 
If you add background to the source, does it not appear if the background is the property of the source? Background seems more to do with observation than with a source.
The source is included in multiple observations, with different backgrounds, what do you do then?
Maybe background could be another source, a "background" source, not part of the given point source.

**Comment by grburgess** [comment]
I've found a solution dealing with just the plugin. I'll push tomorrow. Basically I just create an internal function and assign its parameters as nuisance.

**Comment by grburgess** [comment]
See here:
https://github.com/giacomov/3ML/commit/8a6c0e2bb7e56f5d5f40f14fffb9bf1158266166

**Comment by giacomov** [comment]
Hi,
as a general rule, you probably shouldn't have a discussion and act on it on a weekend... ;-) 

Since the user is likely going to need to iterate the background modeling until he/she finds a good model, I think the most sensible way to handle it is through a separate source. That way you can interact with it using all the tools we already have, and then we you have found a good solution, you use it as background for something else.

Tomorrow I'll think this through a bit more and come up with some actionable idea.

**Comment by grburgess** [comment]
No worries, this is basically what is going on except that it is buried in the plugin. This can be taken further outside, but the interaction will need to take place inside the plugin.

**Comment by grburgess** [comment]
Here are my thoughts and my approach.

With SpectrumLike, we currently sum all the sources associated with the model when computing the likelihood:

```python

    @staticmethod
    def _get_diff_flux_and_integral(likelihood_model):

        n_point_sources = likelihood_model.get_number_of_point_sources()

        # Make a function which will stack all point sources (OGIP do not support spatial dimension)

        def differential_flux(energies):
            fluxes = likelihood_model.get_point_source_fluxes(0, energies)

            # If we have only one point source, this will never be executed
            for i in range(1, n_point_sources):
                fluxes += likelihood_model.get_point_source_fluxes(i, energies)

            return fluxes

        # The following integrates the diffFlux function using Simpson's rule
        # This assume that the intervals e1,e2 are all small, which is guaranteed
        # for any reasonable response matrix, given that e1 and e2 are Monte-Carlo
        # energies. It also assumes that the function is smooth in the interval
        # e1 - e2 and twice-differentiable, again reasonable on small intervals for
        # decent models. It might fail for models with too sharp features, smaller
        # than the size of the monte carlo interval.

        def integral(e1, e2):
            # Simpson's rule

            return (e2 - e1) / 6.0 * (differential_flux(e1)
                                      + 4 * differential_flux((e1 + e2) / 2.0)
                                      + differential_flux(e2))

        return differential_flux, integral

```

This makes sense in the total = bkg+source part of the likelihood, but the conditional background part need to sum only over background sources, and we would need some kind of flag to differentiate. 

I tried a rather ugly approach of building a Source inside the plugin:

```python

 def set_background_model(self, background_shape):

        # we will create a point source for the background
        # with this shape

        bkg_ps = PointSource('background_model',0,0,spectral_shape=background_shape)

        self._background_model = Model(bkg_ps)

        # now extract the free parameters and set them to the nuisance list


        new_nuisance_parameters = self._nuisance_parameters

        for par_name, par in self._background_model.parameters.iteritems():


            new_name = "%s_bkg_%s" %(self._name, par.name)


            new_nuisance_parameters[new_name] = par

        differential_flux, integral = self._get_diff_flux_and_integral(self._background_model)


        self._background_integral_flux = integral

```

The source created here could be backed and created before passing to the class. What I do not like is that this is breaking the idea of sources passed to the model. You could instead add the source to the Model(), and then pass some sort of link that tells the plugin to only use that one for the background.

I took this approach following the idea that the background in the LAT plugin is not really an astromodels  "Source" but an internal property. Could that also be changed?

**Comment by giacomov** [comment]
You are both touching on different issues here. Let me focus for the moment on how to achieve a "functional" background, instead of using the profile likelihoods.

Some thoughts in random order:

* we need to provide a way for the user to iterate over the background fit, and seeing its results (residuals and so on). Indeed, we cannot assume that the user will know every time the function to be used and might need to adjust it.

* the background pertain to an observation, i.e., to a plugin instance as @grburgess implied with his solution. Thus logically the background model does not belong into the Model itself (except for its parameters, that become nuisance parameters). Indeed, a Model is supposed to be an idealization of the sky and do not depend on any instrument

* we have at least two cases: the typical X-ray case, where you have an observed spectrum and a background spectrum both measured directly, and the GBM-type case where you do not really have an independent background measurement, but you need to build it by interpolation. In both case, I would say, the background model should be compared to the background data and its likelihood summed to the total likelihood for the plugin. so that the background model is required also to model the background measurement and cannot go too crazy.

That being said, maybe a workflow would be:

```python

back = XYLike.from_OGIP_background(...)

# or

back = XYLike.from_polynomial_fit(...)

# then fit a function to the background:

fitfunction = Powerlaw() # or whatever
parameters, like_values = xyl.fit(fit_function)
xyl.plot(x_scale='log', y_scale='log')

# Iterate until a good model is found, then:

ogip = OGIPLike(...., background=back)

# continue as usual
```

Note that I use the XYLike plugin as background for multiple reasons:

* in general a background should be modeled in the count space, so no response is necessary
* xylike can handle both Poisson and Gaussian data
* we can use the get_log_like() method of XYLike inside OGIPLike to get the log-likelihood of the background model, and the function to get the expected background flux

What do you think?

**Comment by grburgess** [comment]
I like the idea of having the background come from a plugin. Though, I wonder how biased you can become from fitting the background alone. It is probably mitigated with a good global optimizer and only a detail here.

I am modeling the background in count space and was going to ask you @giacomov if there was any case where that would not apply (it is emphasized in the XSPEC documentation). 

I'm not opposed to XYLike, but SpectrumLike is binned, so maybe HistLike is a better choice? I can think of a few other issues, but in general I believe this approach is the same expect that the information is buried in the plugin instance rather.

So, when the background is coming from another plugin, 

* I will cycle through the plugin's parameters and add then as nuisance so that the fitting process picks them up
* call the background model to get predicted counts 
* use get log like from the other plugin to sum with the other likelihoods

One issue I ran into is renaming parameters. I can name them in the dictionary to be associated with a plugin, but not in the parameter.name itself. @giacomov, what is the *safe* way to do this?

**Comment by volodymyrss** [comment]
Hm, @giacomov, for me, the model is a model of representation of physical reality not just sky. Which is why it might include detector properties, in particular the background.
Otherwise, where do I keep them?
Of course, if they are specific to only one instance of plugin, they can be all hidden, but if not?

**Comment by giacomov** [comment]
@volodymyrss you have a point, but there are different types of background. 

A phenomenological background such as the one obtained from data in a typical X-ray on/off analysis definitely belongs to the data side, since it's applicability never goes beyond that specific data. 

A more astronomical background such as the Galactic diffuse emission could as well belong to the Model, as it is always going to be there whatever instrument you choose to analyze that particular region of the sky.

**Comment by volodymyrss** [comment]
Ok, but where I keep instrumental parameters, which I want to change and fit, and which shared by multiple plugin instances?
Is there a place for an Instrument Model?

**Comment by giacomov** [comment]
@volodymyrss it would help me a lot if you could lay out the problem you are trying to solve. I guess it is an INTEGRAL analysis, can you describe in details how do you need it to work?

Normally nuisance parameters are added to the model by the plugin as "external parameters". Once each plugin has added its own nuisance parameters, you can also link them between each other and obtain what you want. But without details on what you want to achieve, it is not possible for me to understand if this is sufficient or not...

**Comment by grburgess** [comment]
Regarding the 

```python
 XYLike.from_OGIP(....)
``` 
approach. Since these are all going to come from SpectrumLike and need the functionality of energy selection, etc., should we not build the internal background function from SpectrumLike rather than XYLike?

**Comment by grburgess** [comment]
@giacomov one other issue I will run across is the rebinning and channel selection. They should remain the same between background data and total data. One way to solve is with a decorator on all those functions which sets the background plugin to the same selections/bins. This way you only need to add the get_log_like() of the background file. 

The same behavior could be had by calculating the likelihood directly in the total plugin. Just depends on how we want it to look.

**Comment by giacomov** [comment]
@grburgess I am not sure that it is the case. You can have a selection for the background with which you fit the background model, and a different selection for the "source".

For example, in the GBM you might be confident that you know how to model the K-edge in the counts space for the background spectrum (maybe putting a Gaussian there would suffice), however you do not trust the photon response in the middle of it and you exclude that region from the "source" spectrum. In this case you shouldn't exclude the K-edge from the background spectrum, but you should from the source one. Am I right?

**Comment by grburgess** [comment]
Yup :) I was reading https://arxiv.org/abs/1007.1727 on a plane and realized this. It will then be on the user to really know what they are doing. I'll need to do some math and then I'll post my idea

**Comment by volodymyrss** [comment]
@giacomov exactly, I have now tons of external parameters. They work similarly to cons_* things, except evolve with time. There are several needs for this.
But let me focus on one, related for background, since that's what this #227 issue is about.

Background in ISGRI consists of CXB (1), stray light from unmodeled off-axis sources (2), photon background generated in the spacecraft by photons and particles (3), and particles in the detector itself (4).
By default, we use maps averaged over long term, which sum all these things.
It is possible to model these components to a large degree. 
So, following the logic you describe, I would put (1) and (2) in the model, and keep (3) and (4) in the plugin or external properties? Or maybe I put (3) in the model too (some of it is photons after all)? What if 
I want to characterize local particle density, should I than put (4) in the model too? 

I get that there is some physical sky independent of which instrument is used, and reconstructing it is the primary purpose of 3ML.

However, there is also a physical reality of instrument, and it might be hard to separate it from the sky contribution at first. To separate, by fitting, they are treated at the same level, as model.
Which is why I really need to have something like "instrument model" (which contains all kinds of background, and much more). Currently I achieve it with external parameters.

If you allow cons_* parameters, why not treat others in the same way (e.g background normalization, background shape parameters)

To be clear, what I described is not quite standard pipeline.

**Comment by grburgess** [comment]
@volodymyrss I think you can still achieve this through modeling. However, some of these things seem like they should be outside the plugin as external software. Nevertheless, if you can model then, I suppose you can write them in as nuisance parameters. This is totally allowed.


I've now got the first attempt of background modeling implemented in #232 

There is an example added that demonstrates how it is used.
